File: cs/Problems/Day1.cs
namespace aoc24.Problems;

// https://adventofcode.com/2024/day/1
public sealed class Day1 : IProblem<int>
{
    public int Solve(string input) => DistanceBetweenListsOptimized(input);

    private static int DistanceBetweenListsOptimized(ReadOnlySpan<char> input)
    {
        int lineCount = input.Count(InputReader.NewLine) + 1;

        Span<int> left = stackalloc int[lineCount];
        Span<int> right = stackalloc int[lineCount];
        Span<Range> ranges = stackalloc Range[lineCount];

        input.Split(ranges, InputReader.NewLine);

        for (int i = 0; i < lineCount; i++)
        {
            var line = input[ranges[i]];
            int inx = line.IndexOf("   ");

            left[i] = int.Parse(line[..inx]);
            right[i] = int.Parse(line[(inx + 3)..]);
        }

        MemoryExtensions.Sort(left);
        MemoryExtensions.Sort(right);

        int distance = 0;

        for (int i = 0; i < lineCount; i++)
            distance += Math.Abs(left[i] - right[i]);

        return distance;
    }
}

File: go/problems/day1.go
package problems

import (
	"slices"
	"strconv"
	"strings"

	"github.com/Meryovi/hobby-aoc-2024/go/shared"
)

type Day1 struct{}

func (d Day1) Solve(input string) int {
	lines := strings.Split(input, shared.NewLine)
	right := make([]int, len(lines))
	left := make([]int, len(lines))

	for i, line := range strings.Split(input, shared.NewLine) {
		split := strings.Split(line, "   ")
		left[i], _ = strconv.Atoi(split[0])
		right[i], _ = strconv.Atoi(split[1])
	}

	slices.Sort(left)
	slices.Sort(right)

	r := 0

	for i := range left {
		r += IntAbs(left[i] - right[i])
	}

	return r
}

func IntAbs(n int) int {
	if n < 0 {
		return -n
	}
	return n
}
